File: Java/leetcode/editor/en/375.guess-number-higher-or-lower-ii.java
package leetcode.editor.en;/*
 * @lc app=leetcode id=375 lang=java
 *
 * [375] Guess Number Higher or Lower II
 */

// @lc code=start

// Memory Search O(n^3)
class Solution {
    static int N = 202;
    static int[][] memory = new int[N][N];
    
    public int getMoneyAmount(int n) {
        return dfs(1, n);
    }
    
    int dfs(int l, int r){
        if (l >= r) return 0;
        if (memory[l][r] != 0) return memory[l][r];
        int ans = 0x3f3f3f3f;
        for (int x=l; x<=r; x++){
            int cur = Math.max(dfs(l, x-1), dfs(x+1, r)) + x;
            ans = Math.min(ans, cur);
        }
        memory[l][r] = ans;
        return ans;
    }
}

// @lc code=end

File: VSCode_Python/375.guess-number-higher-or-lower-ii.py
#
# @lc app=leetcode id=375 lang=python3
#
# [375] Guess Number Higher or Lower II
#

# @lc code=start
class Solution:
    def getMoneyAmount(self, n: int) -> int:
        infinity = 10 ** 9
        dp = [[0]*(n+1) for _ in range(n+1)]
        for i in range(n-1, 0, -1):
            for j in range(i+1, n+1):
                dp[i][j] = j + dp[i][j-1]
                for k in range(i, j):
                    dp[i][j] = min(dp[i][j], k+max(dp[i][k-1], dp[k+1][j]))
        return dp[1][n]


# @lc code=end

